### PR TITLE
Dev/55 Rejudge

### PR DESCRIPTION
**Pull request by nyngwang**
image
image
image
image

### CODE REVIEW

**Comment by Johnny850807** [comment]
Happy Birthday!

**Comment by timt1028** [comment]
╬╔╦╦═╦═╦═╦╦╗╔═╦╦═╦══╦╦╦═╦═╦╦╗╬
╬║╩║╬║╬║╬║║║║╚║║╬╠╗╔╣╩║║║╬║║║╬
╬║╦║║║╔╣╔╬╗║║╔║║═╣║║║╦║║║║╠╗║╬
╬╚╩╩╩╩╝╚╝╚═╝╚═╩╩╩╝╚╝╚╩╩═╩╩╩═╝╬

**Comment by snowmancc** [comment]
╬╔╦╦═╦═╦═╦╦╗╔═╦╦═╦══╦╦╦═╦═╦╦╗╬
╬║╩║╬║╬║╬║║║║╚║║╬╠╗╔╣╩║║║╬║║║╬
╬║╦║║║╔╣╔╬╗║║╔║║═╣║║║╦║║║║╠╗║╬
╬╚╩╩╩╩╝╚╝╚═╝╚═╩╩╩╝╚╝╚╩╩═╩╩╩═╝╬

**Review by timt1028 — APPROVED** [review]

**Comment by Chaoyuuu** [comment]
image

**Comment by haribo0915** [comment]
╬╔╦╦═╦═╦═╦╦╗╔═╦╦═╦══╦╦╦═╦═╦╦╗╬
╬║╩║╬║╬║╬║║║║╚║║╬╠╗╔╣╩║║║╬║║║╬
╬║╦║║║╔╣╔╬╗║║╔║║═╣║║║╦║║║║╠╗║╬
╬╚╩╩╩╩╝╚╝╚═╝╚═╩╩╩╝╚╝╚╩╩═╩╩╩═╝╬
@Johnny850807 應該給壽星 @nyngwang 多幾個 issue 當生日禮物

**Review by Johnny850807 — COMMENTED** [review]
> ╬╔╦╦═╦═╦═╦╦╗╔═╦╦═╦══╦╦╦═╦═╦╦╗╬
> ╬║╩║╬║╬║╬║║║║╚║║╬╠╗╔╣╩║║║╬║║║╬
> ╬║╦║║║╔╣╔╬╗║║╔║║═╣║║║╦║║║║╠╗║╬
> ╬╚╩╩╩╩╝╚╝╚═╝╚═╩╩╩╝╚╝╚╩╩═╩╩╩═╝╬
> @Johnny850807 應該給壽星 @nyngwang 多幾個 issue 當生日禮物

**Comment by loag20056** [comment]
╬╔╦╦═╦═╦═╦╦╗╔═╦╦═╦══╦╦╦═╦═╦╦╗╬
╬║╩║╬║╬║╬║║║║╚║║╬╠╗╔╣╩║║║╬║║║╬
╬║╦║║║╔╣╔╬╗║║╔║║═╣║║║╦║║║║╠╗║╬
╬╚╩╩╩╩╝╚╝╚═╝╚═╩╩╩╝╚╝╚╩╩═╩╩╩═╝╬

建中王甯我老大 生日快樂歐

**Comment by zseqw123** [comment]
╬╔╦╦═╦═╦═╦╦╗╔═╦╦═╦══╦╦╦═╦═╦╦╗╬
╬║╩║╬║╬║╬║║║║╚║║╬╠╗╔╣╩║║║╬║║║╬
╬║╦║║║╔╣╔╬╗║║╔║║═╣║║║╦║║║║╠╗║╬
╬╚╩╩╩╩╝╚╝╚═╝╚═╩╩╩╝╚╝╚╩╩═╩╩╩═╝╬

File: .idea/workspace.xml [hunk 1]
```diff
@@ -4,8 +4,40 @@
     <option name="autoReloadType" value="SELECTIVE" />
   </component>
   <component name="ChangeListManager">
+<<<<<<< HEAD
+<<<<<<< HEAD
+<<<<<<< HEAD
+<<<<<<< HEAD
```
[the remark this answers is not in the document]
Comment (timt1028) [reply .idea/workspace.xml:10]: WTF

**Comment by Wally5077** [comment]
╬╔╦╦═╦═╦═╦╦╗╔═╦╦═╦══╦╦╦═╦═╦╦╗╬
╬║╩║╬║╬║╬║║║║╚║║╬╠╗╔╣╩║║║╬║║║╬
╬║╦║║║╔╣╔╬╗║║╔║║═╣║║║╦║║║║╠╗║╬
╬╚╩╩╩╩╝╚╝╚═╝╚═╩╩╩╝╚╝╚╩╩═╩╩╩═╝╬
你大寧哥，生日快樂啦

**Comment by ntuhan** [comment]
生日快樂甯哥 什麼時候要下去ㄋ

**Comment by b4735268** [comment]
╬╔╦╦═╦═╦═╦╦╗╔═╦╦═╦══╦╦╦═╦═╦╦╗╬
╬║╩║╬║╬║╬║║║║╚║║╬╠╗╔╣╩║║║╬║║║╬
╬║╦║║║╔╣╔╬╗║║╔║║═╣║║║╦║║║║╠╗║╬
╬╚╩╩╩╩╝╚╝╚═╝╚═╩╩╩╝╚╝╚╩╩═╩╩╩═╝╬
老大，約甲奔阿

File: .idea/workspace.xml (hunk 1 above)
[the remark this answers is not in the document]
Comment (nyngwang) [reply .idea/workspace.xml:10]: 笑死 我自己都不知道有這個

File: src/app/exam/problems/ExamProblems.scss [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+    display: inline;
+    margin: 1rem 0;
+    padding: 5px 20px;
+    cursor: pointer;
+
+    span {
+      color: #FFF;
+      font-weight: bold;
+      font-size: 20px;
+      //line-height: 30px;
```
Comment (Johnny850807) [inline src/app/exam/problems/ExamProblems.scss:25]: remove this

File: src/services/SubmissionService.js [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+export class SubmissionService {
+
+    constructor() {
+        this.axios = axios.create({
+            baseURL: process.env.REACT_APP_PROBLEM_SVC_BASE_URL,
+            timeout: 10000
+        });
+    }
+
+    async submit(problem) {
```
Comment (Johnny850807) [inline src/services/SubmissionService.js:12]: Should be renamed to `rejudge`

and don't implement this method at this moment, the issue created for this #76

File: src/services/ExamService.js [hunk 4]
```diff
@@ -54,6 +54,11 @@ export class ExamService {
         return this.axios.post(`/api/exams/${examId}/problems/${problemId}`, question).then(res => new Question(res.data));
     }
 
+    async deleteExamProblem(problem) {
```
Comment (Johnny850807) [inline src/services/ExamService.js:57]: ` async deleteExamProblem(question) {`

File: src/app/exam/problems/ExamProblems.js [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+        }
+    }];
+
+    const rejudgeProblem = (problemId) => {
+        setShowRejudgeProblemModal(0)
+        setRejudgeProblemId(problemId)
+        submissionService.submit({
+            problemId: problemId,
+            langEnvName: "C",
+            studentId: "r09900000"
```
Comment (Johnny850807) [inline src/app/exam/problems/ExamProblems.js:84]: no, don't hard-code

File: src/services/SubmissionService.js [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+        const {problemId, langEnvName, studentId} = problem
```
Comment (Johnny850807) [inline src/services/SubmissionService.js:13]: wtf is this? `const {problemId, langEnvName, studentId} = problem`

please don't rush on the code.
Ask @snowmancc for the api details (rejudge)

File: src/app/exam/problems/ExamProblems.js [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+        if (!examProblems) {
+            examService.getExam(examId)
+                .then(exam => {
+                    exam.questions.sort((questionA, questionB) => questionA.questionOrder - questionB.questionOrder);
+                    setExamProblems(exam.questions);
+                });
+        } else {
+            examProblems.forEach(problem => {
+                problemService.getProblemById(problem.problemId)
+                    .then(res => setProblemId2Title(prev => prev.set(problem.problemId, res.title)))
```
Comment (Johnny850807) [inline src/app/exam/problems/ExamProblems.js:50]: don't name anything with `2`

File: src/app/exam/problems/ExamProblems.js [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+            })
```
Comment (Johnny850807) [inline src/app/exam/problems/ExamProblems.js:51]: don't iterate on every problem

the get exam overview api has already included the problem's title

**Review by Johnny850807 — CHANGES_REQUESTED** [review]
Review done.

File: src/app/exam/problems/ExamProblems.scss (hunk 2 above)
Comment (Johnny850807) [reply src/app/exam/problems/ExamProblems.scss:25]: @nyngwang unresolved

File: src/app/exam/problems/ExamProblems.js [hunk 9]
```diff
@@ -0,0 +1,171 @@
...
+        });
+    }, [examId])
+
+    useEffect(() => {
+        if (!examProblems) {
+            fetchExam()
+        }
+    }, [examProblems, fetchExam]);
+
+    const editProblem = (problemId) => {
```
Comment (Johnny850807) [inline src/app/exam/problems/ExamProblems.js:45]: rename `editProblem` to `editQuestion`

File: src/app/exam/problems/ExamProblems.js [hunk 10]
```diff
@@ -0,0 +1,171 @@
...
+        submissionService.rejudge({
+            examId: examId,
+            problemId: problemId
+        }).then(res => {
+            console.log(res)
+        })
+        setTimeout(() =>
+        {
+            setRejudgeProblemId(0)
+        }, 5000)
```
Comment (Johnny850807) [inline src/app/exam/problems/ExamProblems.js:87]: why is the need for `setTimeout`? 

do this in the "then" section 

and what does `0` mean?

File: src/app/exam/problems/ExamProblems.js [hunk 11]
```diff
@@ -0,0 +1,171 @@
...
+            setShowRejudgeProblemModal(examProblem.problemId)
+        }
+    }, {
+        name: "Delete",
+        dangerous: true,
+        onClick: () => {
+            examService.deleteExamProblem({
+                examId: examId,
+                problemId: examProblem
+            }).then(res => console.log("deleted", res))
```
Comment (Johnny850807) [inline src/app/exam/problems/ExamProblems.js:71]: give a more clearer comment, this is unclear when we debug from a ton of logs

File: src/app/exam/problems/ExamProblems.js [hunk 12]
```diff
@@ -0,0 +1,171 @@
...
+
+const ExamProblems = ({ exams }) => {
+    const { url: currentURL } = useRouteMatch()
+    const { examId } = useParams()
+    const currentExamName = exams.find(exam => exam.id === parseInt(examId))?.name
+    const [examProblems, setExamProblems] = useState(null);
+
+    const [showAddProblemModal, setShowAddProblemModal] = useState(false);
+    const [showRejudgeProblemModal, setShowRejudgeProblemModal] = useState(0);
+    const [rejudgeProblemId, setRejudgeProblemId] = useState(0)
```
Comment (Johnny850807) [inline src/app/exam/problems/ExamProblems.js:27]: what does `0` mean?

Declare a constant to describe it.
More usually, it should be`-1`?

File: src/app/exam/problems/ExamProblems.js [hunk 13]
```diff
@@ -0,0 +1,171 @@
...
+const toCharacterIndex = i => {
+    return String.fromCharCode(i + 65);
+}
+
+
+const ExamProblems = ({ exams }) => {
+    const { url: currentURL } = useRouteMatch()
+    const { examId } = useParams()
+    const currentExamName = exams.find(exam => exam.id === parseInt(examId))?.name
+    const [examProblems, setExamProblems] = useState(null);
```
Comment (Johnny850807) [inline src/app/exam/problems/ExamProblems.js:23]: Don't understand, why all exams are passed in.

Just pass the current exam in.

**Review by Johnny850807 — CHANGES_REQUESTED** [review]
Review done

File: src/app/exam/problems/ExamQuestions.js [hunk 14]
```diff
@@ -0,0 +1,166 @@
...
+    }];
+
+    const rejudgeQuestion = (questionId) => {
+        setRejudgeQuestionId(questionId)
+        submissionService.rejudge({examId, questionId})
+            .then(res => {
+                console.log("Calling Rejudge API: and get result:", res)
+                setRejudgeQuestionId(NOT_SET)
+            })
+    }
```
Comment (Johnny850807) [inline src/app/exam/problems/ExamQuestions.js:85]: `questionId` -> `problemId`

**question's id = (examId, problemId)**

File: src/app/commons/IndexBanner/InPageNavigationBar.js [hunk 15]
```diff
@@ -20,13 +20,16 @@ function link(currentPathName, to, name, Icon) {
  *              (3) Icon, the Icon on the tab
  */
 
-const InPageNavigationBar = function ({currentURL, path, tabContents}) {
+const InPageNavigationBar = function ({currentURL, path, tabContents, events}) {
     const secondSlashPos = currentURL.substr(1).indexOf('/')
     const to = currentURL.substr(0, 1+secondSlashPos)
+    const { firstParamResetState } = events
```
Comment (Johnny850807) [inline src/app/commons/IndexBanner/InPageNavigationBar.js:26]: add refactor TODO

**Review by Johnny850807 — CHANGES_REQUESTED** [review]
Review done

**Review by Johnny850807 — APPROVED** [review]
ＬＧＴＭ

**Pull request merged**